common/src/Microsoft.Azure.IIoT.Agent.Framework/src/Agent/Default/Worker.cs

```diff
@@ -94,22 +87,26 @@ public class Worker : IWorker, IDisposable {
 
         /// <inheritdoc/>
         public async Task StopAsync() {
-            await _lock.WaitAsync().ConfigureAwait(false);
```

Comment:
Given that `StartAsync()` and `GetStatus()` rely on the `_lock` to make sure that all operations are synchronous I would suggest to revert back the `_lock` related changes here. With the current changes we open-up opportunities for fast call to `StopAsync()` to not perform intended operations if there was a `StartAsynch()` executing at the same time. I understand that these kind of rapid calls might be exceptional cases, but the whole point of having `_lock` here seem to be to safeguard against that kind of situations.

Reply:
The other possible error that I can see is that we might assign `_jobProcess` to `null` (on line `105`) at the same time as `GetStatus()` was evaluating `_jobProcess?.Status` operation. So the `?` check might see that `_jobProcess` is not null, but when it tries to access its `Status` member the `_jobProcess` might already be null and thus cause a `NullReferenceException`.

Reply:
nope, the reason for this change was a super ugly deadlock and exactly because of the lock on the `Status` in `SendHeartbeatWithoutResetTimerAsync`